### ISSUE TITLE
DebugPool not visible in hierarchy

### ISSUE THREAD

**Issue by sschmid** [sschmid/Entitas-CSharp#108]
Sometimes DebugPools don't appear in the hierarchy when entering play-mode.
It seems to me that this is a Unity bug.

DebugPools are only created when UNITY_EDITOR and Application.isPlaying.

Every time you enter play-mode both UNITY_EDITOR and Application.isPlaying is true, that's why you see the pools in the hierarchy.

However, if you change some code and go back to Unity and hit play while Unity is still compiling, it will enter play-mode as soon as it finished compiling, but suddenly Application.isPlaying is false...

You can test it yourself

```csharp
using UnityEditor;
using UnityEditor.Callbacks;
using UnityEngine;

public class PlayBugController : MonoBehaviour {

    [DidReloadScripts]
    public static void OnDidReloadScripts() {
        Debug.Log("OnDidReloadScripts");
        Debug.Log("isPlaying: " + EditorApplication.isPlaying);
        Debug.Log("isPlayingOrWillChangePlaymode: " + EditorApplication.isPlayingOrWillChangePlaymode);
    }
}
```

Try: Hit play
Try Hit play while compiling

-> different results

I'll file a bug. If you have a workaround, let me know

**Comment by sschmid** [comment]
Test project with the bug

[IsPlayingBug.zip](https://github.com/sschmid/Entitas-CSharp/files/262134/IsPlayingBug.zip)

**Comment by sschmid** [comment]
Filed a bug
https://support.unity3d.com/hc/en-us/requests/49030

**Comment by sschmid** [comment]
Current quick fix:
hit stop and play again :) or wait until Unity finished compling

**Comment by vladpazych** [comment]
Or you can do this:
```csharp
static Pool _pool;
static GameObject _poolObserverContainer = null;

public static Pool pool {
	get {
		if (_pool == null) {
			_pool = new Pool (ComponentIds.TotalComponents, 0, new PoolMetaData ("Pool", ComponentIds.componentNames, ComponentIds.componentTypes));
		}

		#if (!ENTITAS_DISABLE_VISUAL_DEBUGGING && UNITY_EDITOR)
		if (_poolObserverContainer == null && UnityEditor.EditorApplication.isPlaying) {
			var poolObserver = new Entitas.Unity.VisualDebugging.PoolObserver (_pool);
			_poolObserverContainer = poolObserver.entitiesContainer;
			UnityEngine.Object.DontDestroyOnLoad (_poolObserverContainer);
		}
		#endif

		return _pool;
	}
}
```

**Comment by JamesMcMahon** [comment]
I'm seeing an issue with 0.31.2 where I see the pool game object in my scene when NOT in play mode. Is that related to this issue?

**Comment by sschmid** [comment]
That is exactly this issue. Happens when entering play-mode while still compiling.
After stopping and entering play-mode, it should be visible.

**Comment by JamesMcMahon** [comment]
I see the follow exception when it happens:

```
InvalidOperationException: The following game object is invoking the DontDestroyOnLoad method: Pool (0 entities, 0 reusable, 0 groups). Notice that DontDestroyOnLoad can only be used in play mode and, as such, cannot be part of an editor script.
Pools.get_pool ()
Pools.get_allPools ()
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture)
System.Reflection.MonoProperty.GetValue (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] index, System.Globalization.CultureInfo culture)
System.Reflection.MonoProperty.GetValue (System.Object obj, System.Object[] index)
Entitas.Unity.Serialization.Blueprints.BinaryBlueprintInspector.findAllPools ()
Entitas.Unity.Serialization.Blueprints.BinaryBlueprintInspector.UpdateAllBinaryBlueprints ()
```

From `pool.cs`, line 27:
```csharp
public static Pool pool {
    get {
        if (_pool == null) {
            _pool = new Pool(ComponentIds.TotalComponents, 0, new PoolMetaData("Pool", ComponentIds.componentNames, ComponentIds.componentTypes));
            #if (!ENTITAS_DISABLE_VISUAL_DEBUGGING && UNITY_EDITOR)
            var poolObserver = new Entitas.Unity.VisualDebugging.PoolObserver(_pool);
            UnityEngine.Object.DontDestroyOnLoad(poolObserver.entitiesContainer);
            #endif
        }

        return _pool;
    }
}
```

**Comment by sschmid** [comment]
Generate again, that looks like old code

**Comment by JamesMcMahon** [comment]
Oh yup totally, forgot to regenerate when I upgraded.

Code is now

```csharp
public static Pool pool {
    get {
        if (_pool == null) {
            _pool = new Pool(ComponentIds.TotalComponents, 0, new PoolMetaData("Pool", ComponentIds.componentNames, ComponentIds.componentTypes));
            #if (!ENTITAS_DISABLE_VISUAL_DEBUGGING && UNITY_EDITOR)
            if (UnityEngine.Application.isPlaying) {
                var poolObserver = new Entitas.Unity.VisualDebugging.PoolObserver(_pool);
                UnityEngine.Object.DontDestroyOnLoad(poolObserver.entitiesContainer);
            }
            #endif
        }

        return _pool;
    }
}
```

**Comment by sschmid** [comment]
:) No worries

**Comment by sschmid** [comment]
Fyi, Unity was able to reproduce the bug is currently working on a fix.

**Comment by sschmid** [comment]
Still open, but seems like there has bee some work done...
<img width="970" alt="editorapplication isplayingorwillchangeplaymode" src="https://cloud.githubusercontent.com/assets/233700/16151889/7190f0ce-349f-11e6-9e3f-80c1fb139f15.png">

**Comment by bddckr** [comment]
This is fixed in Unity 5.4.0f2 (release candidate 2), at least on my end. Can't find anything related to it in the release notes, though.
